File: src/waypoint/sources/local_cache.py
"""Local, content-addressed driver cache.

This is the one source that runs fully today (no network dependency), and
it's the direct fix for SDI's "one monolithic 20-60GB blob" model: a
technician adds drivers here incrementally, one at a time, keyed by their
SHA-256 hash rather than by filename/folder convention. Re-scans across
many machines dedupe automatically because the key *is* the content hash.

Layout on disk:
    <cache_root>/
      manifest.json          # list of DriverCandidate records (as dicts)
      blobs/<sha256>/        # the actual driver package contents
"""

from __future__ import annotations

import hashlib
import json
import shutil
from dataclasses import asdict
from datetime import date
from pathlib import Path

from waypoint.core.models import DriverCandidate, SignatureType


class LocalCacheSource:
    source_id = "local_cache"

    def __init__(self, cache_root: str) -> None:
        self.cache_root = Path(cache_root)
        self.cache_root.mkdir(parents=True, exist_ok=True)
        (self.cache_root / "blobs").mkdir(exist_ok=True)
        self._manifest_path = self.cache_root / "manifest.json"
        if not self._manifest_path.exists():
            self._manifest_path.write_text("[]")

    def _load_manifest(self) -> list[dict]:
        return json.loads(self._manifest_path.read_text())

    def _save_manifest(self, entries: list[dict]) -> None:
        self._manifest_path.write_text(json.dumps(entries, indent=2, default=str))

    def add_package(
        self,
        source_file: str,
        *,
        hwid: str,
        class_guid: str,
        version: str,
        driver_date: date | None,
        publisher: str,
        signature_type: SignatureType,
    ) -> DriverCandidate:
        """Register a driver package a technician has vetted, keyed by its
        content hash. This is the explicit, opt-in alternative to trusting
        an opaque third-party archive.
        """
        src_path = Path(source_file)
        sha256 = _hash_file(src_path)
        blob_dir = self.cache_root / "blobs" / sha256
        if not blob_dir.exists():
            blob_dir.mkdir(parents=True)
            shutil.copy2(src_path, blob_dir / src_path.name)

        candidate = DriverCandidate(
            hwid=hwid,
            class_guid=class_guid,
            version=version,
            driver_date=driver_date,
            publisher=publisher,
            signature_type=signature_type,
            sha256=sha256,
            size_bytes=src_path.stat().st_size,
            source_id=self.source_id,
            source_url=str(src_path),
            download_uri=str(blob_dir / src_path.name),
        )
        entries = self._load_manifest()
        entries.append(asdict(candidate))
        self._save_manifest(entries)
        return candidate

    def search(self, hwids: list[str]) -> list[DriverCandidate]:
        entries = self._load_manifest()
        hwid_set = set(hwids)
        results = []
        for entry in entries:
            if entry["hwid"] in hwid_set:
                entry = dict(entry)
                entry["signature_type"] = SignatureType(entry["signature_type"])
                if entry.get("driver_date"):
                    entry["driver_date"] = date.fromisoformat(entry["driver_date"])
                results.append(DriverCandidate(**entry))
        return results

    def fetch(self, candidate: DriverCandidate, dest_dir: str) -> str:
        src = Path(candidate.download_uri)
        if _hash_file(src) != candidate.sha256:
            raise ValueError(
                f"Hash mismatch for {src}: cache is corrupt or candidate metadata is stale."
            )
        dest_path = Path(dest_dir) / src.name
        Path(dest_dir).mkdir(parents=True, exist_ok=True)
        shutil.copy2(src, dest_path)
        return str(dest_path)


def _hash_file(path: Path) -> str:
    hasher = hashlib.sha256()
    with open(path, "rb") as f:
        for chunk in iter(lambda: f.read(1 << 20), b""):
            hasher.update(chunk)
    return hasher.hexdigest()

File: rust/crates/sources/src/local_cache.rs
//! Local, content-addressed driver cache. Direct port of
//! `sources/local_cache.py`.
//!
//! This is the one source that runs fully today (no network dependency),
//! and it's the direct fix for SDI's "one monolithic 20-60GB blob" model:
//! a technician adds drivers here incrementally, one at a time, keyed by
//! their SHA-256 hash rather than by filename/folder convention. Re-scans
//! across many machines dedupe automatically because the key *is* the
//! content hash.
//!
//! Layout on disk:
//! ```text
//! <cache_root>/
//!   manifest.json          # list of DriverCandidate records
//!   blobs/<sha256>/        # the actual driver package contents
//! ```

use std::fs;
use std::io::Read;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use waypoint_core::{DriverCandidate, SignatureType};

use crate::base::DriverSource;

pub struct LocalCacheSource {
    pub cache_root: PathBuf,
    manifest_path: PathBuf,
}

/// Everything `DriverCandidate` needs for on-disk (de)serialization, kept
/// as its own struct rather than deriving directly on `DriverCandidate` so
/// `waypoint-core` doesn't have to know about this source's manifest
/// format.
#[derive(Serialize, Deserialize)]
struct ManifestEntry {
    hwid: String,
    class_guid: String,
    version: String,
    driver_date: Option<time::Date>,
    publisher: String,
    signature_type: SignatureType,
    sha256: String,
    size_bytes: u64,
    source_id: String,
    source_url: String,
    download_uri: String,
}

impl From<&DriverCandidate> for ManifestEntry {
    fn from(c: &DriverCandidate) -> Self {
        ManifestEntry {
            hwid: c.hwid.clone(),
            class_guid: c.class_guid.clone(),
            version: c.version.clone(),
            driver_date: c.driver_date,
            publisher: c.publisher.clone(),
            signature_type: c.signature_type,
            sha256: c.sha256.clone(),
            size_bytes: c.size_bytes,
            source_id: c.source_id.clone(),
            source_url: c.source_url.clone(),
            download_uri: c.download_uri.clone(),
        }
    }
}

impl From<ManifestEntry> for DriverCandidate {
    fn from(e: ManifestEntry) -> Self {
        DriverCandidate {
            hwid: e.hwid,
            class_guid: e.class_guid,
            version: e.version,
            driver_date: e.driver_date,
            publisher: e.publisher,
            signature_type: e.signature_type,
            sha256: e.sha256,
            size_bytes: e.size_bytes,
            source_id: e.source_id,
            source_url: e.source_url,
            download_uri: e.download_uri,
        }
    }
}

impl LocalCacheSource {
    pub fn new(cache_root: impl AsRef<Path>) -> Result<Self, String> {
        let cache_root = cache_root.as_ref().to_path_buf();
        fs::create_dir_all(&cache_root).map_err(|e| e.to_string())?;
        fs::create_dir_all(cache_root.join("blobs")).map_err(|e| e.to_string())?;
        let manifest_path = cache_root.join("manifest.json");
        if !manifest_path.exists() {
            fs::write(&manifest_path, "[]").map_err(|e| e.to_string())?;
        }
        Ok(Self {
            cache_root,
            manifest_path,
        })
    }

    fn load_manifest(&self) -> Result<Vec<ManifestEntry>, String> {
        let raw = fs::read_to_string(&self.manifest_path).map_err(|e| e.to_string())?;
        serde_json::from_str(&raw).map_err(|e| e.to_string())
    }

    fn save_manifest(&self, entries: &[ManifestEntry]) -> Result<(), String> {
        let raw = serde_json::to_string_pretty(entries).map_err(|e| e.to_string())?;
        fs::write(&self.manifest_path, raw).map_err(|e| e.to_string())
    }

    /// Register a driver package a technician has vetted, keyed by its
    /// content hash. This is the explicit, opt-in alternative to trusting
    /// an opaque third-party archive.
    #[allow(clippy::too_many_arguments)]
    pub fn add_package(
        &self,
        source_file: impl AsRef<Path>,
        hwid: &str,
        class_guid: &str,
        version: &str,
        driver_date: Option<time::Date>,
        publisher: &str,
        signature_type: SignatureType,
    ) -> Result<DriverCandidate, String> {
        let src_path = source_file.as_ref();
        let sha256 = hash_file(src_path)?;
        let blob_dir = self.cache_root.join("blobs").join(&sha256);
        if !blob_dir.exists() {
            fs::create_dir_all(&blob_dir).map_err(|e| e.to_string())?;
            let file_name = src_path
                .file_name()
                .ok_or_else(|| "source_file has no file name".to_string())?;
            fs::copy(src_path, blob_dir.join(file_name)).map_err(|e| e.to_string())?;
        }
        let size_bytes = fs::metadata(src_path).map_err(|e| e.to_string())?.len();
        let file_name = src_path
            .file_name()
            .ok_or_else(|| "source_file has no file name".to_string())?;

        let candidate = DriverCandidate {
            hwid: hwid.to_string(),
            class_guid: class_guid.to_string(),
            version: version.to_string(),
            driver_date,
            publisher: publisher.to_string(),
            signature_type,
            sha256: sha256.clone(),
            size_bytes,
            source_id: self.source_id().to_string(),
            source_url: src_path.to_string_lossy().to_string(),
            download_uri: blob_dir.join(file_name).to_string_lossy().to_string(),
        };

        let mut entries = self.load_manifest()?;
        entries.push(ManifestEntry::from(&candidate));
        self.save_manifest(&entries)?;
        Ok(candidate)
    }
}

impl DriverSource for LocalCacheSource {
    fn source_id(&self) -> &str {
        "local_cache"
    }

    fn search(&self, hwids: &[String]) -> Result<Vec<DriverCandidate>, String> {
        let entries = self.load_manifest()?;
        let hwid_set: std::collections::HashSet<&String> = hwids.iter().collect();
        Ok(entries
            .into_iter()
            .filter(|e| hwid_set.contains(&e.hwid))
            .map(DriverCandidate::from)
            .collect())
    }

    fn fetch(&self, candidate: &DriverCandidate, dest_dir: &str) -> Result<String, String> {
        let src = Path::new(&candidate.download_uri);
        let actual_hash = hash_file(src)?;
        if actual_hash != candidate.sha256 {
            return Err(format!(
                "Hash mismatch for {}: cache is corrupt or candidate metadata is stale.",
                src.display()
            ));
        }
        let dest_dir = Path::new(dest_dir);
        fs::create_dir_all(dest_dir).map_err(|e| e.to_string())?;
        let file_name = src
            .file_name()
            .ok_or_else(|| "download_uri has no file name".to_string())?;
        let dest_path = dest_dir.join(file_name);
        fs::copy(src, &dest_path).map_err(|e| e.to_string())?;
        Ok(dest_path.to_string_lossy().to_string())
    }
}

fn hash_file(path: &Path) -> Result<String, String> {
    let mut file = fs::File::open(path).map_err(|e| e.to_string())?;
    let mut hasher = Sha256::new();
    let mut buf = [0u8; 1 << 20];
    loop {
        let n = file.read(&mut buf).map_err(|e| e.to_string())?;
        if n == 0 {
            break;
        }
        hasher.update(&buf[..n]);
    }
    Ok(format!("{:x}", hasher.finalize()))
}
